packages/slate-locales-plugin/index.js

```diff
@@ -0,0 +1,183 @@
+      compilation.assets[outputKey] = new ConcatSource(
+        new RawSource(
+          fs.readFileSync(path.resolve(directoryPath, 'template.liquid')),
+        ),
+        new RawSource(
+          `{% schema %}\n${fs.readFileSync(
+            path.resolve(directoryPath, 'schema.json'),
+          )}{% endschema %}`,
+        ),
+      );
+    } else {
+      compilation.assets[outputKey] = new RawSource(
+        fs.readFileSync(path.resolve(directoryPath, 'template.liquid')),
+      );
+    }
+  } else {
+    const liquidContent = await fs.readFile(
+      path.resolve(directoryPath, liquidDirent.name),
+      'utf8',
+    );
+    const outputKey = `../sections${
+      directoryPath.split('/sections')[1] === ''
+        ? `/${liquidDirent.name}`
+        : `${directoryPath.split('/sections')[1]}/${liquidDirent.name}`
+    }`;
+    if (
+      liquidContent.match(/{%\s*schema\s*%}([\s\S]+){%\s*endschema\s*%}/gim) ||
+      !dirents.find(
+        (obj) => obj.name === liquidDirent.name.replace('.liquid', '.json'),
+      )
+    ) {
+      compilation.assets[outputKey] = new RawSource(liquidContent);
+    } else {
+      compilation.assets[outputKey] = new ConcatSource(
+        new RawSource(liquidContent),
+        new RawSource(
+          `{% schema %}\n${fs.readFileSync(
+            path.resolve(
+              directoryPath,
+              liquidDirent.name.replace('.liquid', '.json'),
+            ),
+          )}{% endschema %}`,
+        ),
+      );
+    }
+  }
+}
+
+async function traverseDirectory(dirent, directoryPath, compilation) {
+  const dirents = await fs.readdir(path.resolve(directoryPath, dirent.name), {
+    withFileTypes: true,
+  });
+  dirents.forEach((subdirent) => {
+    if (subdirent.isDirectory() && subdirent.name !== 'locales') {
+      traverseDirectory(
+        subdirent,
+        path.resolve(directoryPath, dirent.name),
+        compilation,
+      );
+    } else if (
+      subdirent.isFile() &&
+      path.extname(subdirent.name) === '.liquid'
+    ) {
+      getSchema(
+        subdirent,
+        path.resolve(directoryPath, dirent.name),
+        compilation,
+        dirents,
+      );
+    }
+  });
+}
+
+function getLocalizedValues(key, mainSchema, localizedSchema) {
+  // eslint-disable-next-line
+  let combinedTranslationsObject = {};
+  // eslint-disable-next-line
+  for (const i in localizedSchema) {
+    combinedTranslationsObject[i] = _.get(localizedSchema[i], key);
+  }
+
+  return combinedTranslationsObject;
+}
+
+function createMainSchema(localizedSchema, mainSchemaPath) {
+  function traverse(obj) {
+    for (const i in obj) {
+      if (typeof obj[i].t === 'string') {
+        obj[i] = getLocalizedValues(obj[i].t, obj, localizedSchema);
+      } else if (typeof obj[i] === 'object') {
+        traverse(obj[i]);
+      }
+    }
+  }
+  // eslint-disable-next-line
+  let mainSchema = JSON.parse(fs.readFileSync(mainSchemaPath, 'utf-8'));
+  traverse(mainSchema);
+  return mainSchema;
```

Comment:
Would be cleaner if `traverse()` returned `obj` so you could just have `return traverse()`.